### PR TITLE
RavenDB-22717 - serialize dynamic index SupportDuplicateKeys field

### PR DESCRIPTION
**Pull request by garayx**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22717

### Additional description

small fix to add not persisted field in dynamic index schema

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [x] Ensured. 
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by garayx** [comment]
test this in debug please

**Review by ppekrol — APPROVED** [review]

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 1]
```diff
@@ -362,8 +376,25 @@ public static DynamicKeyIndexDef ReadFrom(ByteStringContext context, ref TableVa
                 // read if this index has OnEntryChanged
                 currentPtr = input.Read(5, out currentSize);
                 var hasIndexEntryChangedDelegate = Convert.ToBoolean(*currentPtr);
-                if (hasIndexEntryChangedDelegate == false) 
-                    return indexDef;
+                if (hasIndexEntryChangedDelegate == false)
+                {
+                    if (input.Count == 6)
```
Comment (arekpalinski) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: Is is for backward compatibility purposes? If yes, then please add a comment about that

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 2]
```diff
@@ -362,8 +376,25 @@ public static DynamicKeyIndexDef ReadFrom(ByteStringContext context, ref TableVa
...
+                    {
+                        indexDef.SupportDuplicateKeys = false;
+                        return indexDef;
+                    }
+
+                    if (input.Count == 7)
```
Comment (arekpalinski) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: Can we have some const for those magic numbers (`== 6`, `== 7` etc)?

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 3]
```diff
@@ -362,8 +376,25 @@ public static DynamicKeyIndexDef ReadFrom(ByteStringContext context, ref TableVa
...
+                    {
+                        // read if this index has SupportDuplicateKeys
+                        currentPtr = input.Read(6, out currentSize);
```
Comment (arekpalinski) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: I see that in this file have hard coded indexes to read from. Like `6` here. 

While in other places we use enums so we have:

```
tvr.Read((int)CountersTable.CounterKey, out var size);
```
or
```
LowerId = TableValueToString(context, (int)DocumentsTable.LowerId, ref tvr),
```
This is really helpful when reading or debugging the code. Can we have something like that here too?

**Review by arekpalinski — COMMENTED** [review]
Can you please provide more description what was the actual issue? I talked to Pawel and he told me that it's about compaction but there no mention of that either here or in the issue on YT. Also the stacktrace would be helpful

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 4]
```diff
@@ -291,6 +291,13 @@ public byte[] Serialize()
                                 serializer.CopyTo(destination);
                             }
 
+                            serializer.Add(SupportDuplicateKeys);
+                            serialized = new byte[serializer.Size];
+                            fixed (byte* destination = serialized)
+                            {
+                                serializer.CopyTo(destination);
+                            }
+
```
Comment (aviv86) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: no need to allocate and copy to destination twice 
we can remove the first occurrence of this:
```
	serialized = new byte[serializer.Size];

	fixed (byte* destination = serialized)
	{
		serializer.CopyTo(destination);
	}
```

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 5]
```diff
@@ -317,6 +324,13 @@ public byte[] Serialize()
                                     serializer.CopyTo(destination);
                                 }
 
+                                serializer.Add(SupportDuplicateKeys);
+                                serialized = new byte[serializer.Size];
+                                fixed (byte* destination = serialized)
+                                {
+                                    serializer.CopyTo(destination);
+                                }
+
```
Comment (aviv86) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: same here, no need to copy twice

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 6]
```diff
@@ -433,6 +478,11 @@ public void EnsureIdentical(DynamicKeyIndexDef actual)
                         $"Expected index {Name} to have {nameof(OnEntryChanged)}.Method.DeclaringType='{OnEntryChanged?.Method.DeclaringType}', " +
                         $"got {nameof(GenerateKey)}.Method.DeclaringType='{actual.OnEntryChanged?.Method.DeclaringType}' instead",
                         nameof(actual));
+
+                if (SupportDuplicateKeys != actual.SupportDuplicateKeys)
+                    throw new ArgumentException(
+                        $"Expected index {Name} to be have SupportDuplicateKeys='{SupportDuplicateKeys}', got SupportDuplicateKeys='{actual.SupportDuplicateKeys}' instead",
```
Comment (aviv86) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]:
```suggestion
                        $"Expected index {Name} to have SupportDuplicateKeys='{SupportDuplicateKeys}', got SupportDuplicateKeys='{actual.SupportDuplicateKeys}' instead",
```

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs (hunk 4 above)
Comment (garayx) [reply src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: done

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs (hunk 1 above)
Comment (garayx) [reply src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: done

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs (hunk 3 above)
Comment (garayx) [reply src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: done

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs (hunk 2 above)
Comment (garayx) [reply src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: done

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs (hunk 6 above)
Comment (garayx) [reply src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: done

File: src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs [hunk 7]
```diff
@@ -324,22 +326,45 @@ public byte[] Serialize()
                 }
             }
 
+            private const int SupportDuplicateKeysSmallTableIndex = 6;
+            private enum DynamicIndexTable
+            {
+                Type,
```
Comment (ppekrol) [inline src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: it is a good practice, when value cannot change to not rely on order, but to specify the value explicitly e.g. Type = 0,
IsGlobal = 1

Comment (garayx) [reply src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs]: done

**Review by aviv86 — APPROVED** [review]

**Comment by arekpalinski** [comment]
> Can you please provide more description what was the actual issue? I talked to Pawel and he told me that it's about compaction but there no mention of that either here or in the issue on YT. Also the stacktrace would be helpful

What about that?

**Review by arekpalinski — APPROVED** [review]
The changes LGTM

**Comment by garayx** [comment]
> > Can you please provide more description what was the actual issue? I talked to Pawel and he told me that it's about compaction but there no mention of that either here or in the issue on YT. Also the stacktrace would be helpful
> 
> What about that?

https://issues.hibernatingrhinos.com/issue/RavenDB-22717/serialize-dynamic-index-SupportDuplicateKeys-field#focus=Change-67-1054971.0-0.pinned

**Pull request merged**